content/blog/2022/ci-cd-improvements/index.md

```diff
@@ -0,0 +1,104 @@
+  - Inspect the list of files modified in a Pull Request.
+  - From this list, calculate the name of a hub that required an upgrade, and the name of its respective cluster.
+  - Trigger a GitHub action that would deploy changes in-parallel for each cluster/hub pair.
+
+In addition to these structural and code changes, we also developed new GitHub Actions that control the entire process.
+
+### A GitHub Actions workflow for upgrading our JupyterHubs
+
+We defined a new GitHub Action that carries out the logic described above.
+These are all defined in [this `deploy-hubs.yaml` configuration file](https://github.com/2i2c-org/infrastructure/blob/master/.github/workflows/deploy-hubs.yaml).
+Here are the major steps in this action, and what each does:
+
+1. `generate-jobs`: Generate a list of clusters/hubs that must be upgraded, given the files that are changed in a Pull Request.
+   - Evaluate an input list of added/modified files in a PR
+   - Decide if the added/modified files warrant an upgrade of a hub
+   - Generate a list of hubs and clusters that require upgrades, and some extra details:
+     - Does the support chart that is deployed to the cluster also need an upgrade?
+     - Does a staging hub on this cluster require an upgrade?
+
+   This produced two outputs to be used in subsequent steps:
+
+   - A **human-readable table** including information on _why_ a given deployment requires an upgrade (using the excellent [Rich library](https://github.com/Textualize/rich)).
+   - A **JSON output** that can be interpreted by GitHub Actions as a set of matrix jobs to run.
+
+   % TODO: Can we add an image of this nice table in the post?
+
+2. `upgrade-support-and-staging`: Update the support and staging Helm charts on each cluster. These are "shared infrastructure" Helm charts that control services that are shared across all hubs.
+   - Accepts the JSON list described above to determine what to do next
+   - Parallelises over clusters
+   - Upgrades the support chart of each if required
+   - Upgrades a staging hub for the cluster if required (for canary deployments, this is always required if at least one production hub is to be upgraded on the cluster)
+3. `filter-generate-jobs`: Allows us to treat the support / staging hubs as canary deployments for all the production hubs on a cluster.
+   - If a staging/support hub deploy fails, removes any jobs for the corresponding cluster.
+   - Allows production deploys to continue on _other clusters_.
+4. `upgrade-prod-hubs`: Deploy updates to each production hub.
+   - Accepts the JSON list described above to determine what to do next
+   - Parallelises over each production hub that requires an upgrade
+   - Deploy the relevant changes to that hub
+
+## Concluding Remarks
+
+We think that this is a nicely balance of infrastructure complexity and flexibility.
+It allows us to separate the configuration of each hub and cluster, which makes each more maintainable.
```

Comment:
Do we also want to mention that separating config like this also improves the pathway to R2R since communities can easily find and extract their hub's config?